Add data exploration duplication entry

shamblonaut opened the pull request:
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #4806 

<!-- Concisely describe what the pull request does. -->
This PR adds a "Duplicate Exploration" entry to the context menu for data explorations listed in the schema page.

**Technical details**
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->
It works by adding the existing data exploration to the backend again with a different name. This effectively duplicates the data exploration as the new exploration contains all the data of the previous one except the name.

**Demo**

https://github.com/user-attachments/assets/7e714d45-ed77-44ab-b095-fa0ca0011d20



## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

shamblonaut committed:
Add data exploration duplication entry

Creates a "Duplicate Exploration" entry to the context menu for data
explorations shown in the schema page. It works by saving the existing
exploration to the backend with a different name.

pavish committed:
Merge branch 'develop' into feat-4806-data-explorer-duplicate

pavish reviewed (changes requested):
@shamblonaut Thanks for the PR!

This works mostly well and the code looks good. I'd like the following scenario to be handled as well before we can merge this PR.

Scenario:
- Duplicate an exploration `a`. A new exploration called `Duplicate of a` is created.
- Duplicate the same exploration `a` again. This leads to an error since there's an existing exploration of the same name `Duplicate of a`.
- I'd like the name to be `Duplicate of a (1)` in this case and no error to be thrown.

The naming conflicts should be resolved by checking all existing explorations, and naming the duplicate exploration as `${name} ${i}`, where `i` starts from 1 and is incremented until there is no exploration with the same name.

shamblonaut committed:
Merge branch 'mathesar-foundation:develop' into feat-4806-data-explorer-duplicate

shamblonaut committed:
Apply numeric suffix to exploration duplicates

If an existing duplicate exists when duplicating an exploration that has
not yet been renamed, the new duplicate is saved with a bracketed number
suffix that increments based on existing duplicate names

shamblonaut committed:
Fix type error

suffixNumbers still had the type (number | null)[] after filter, so a
type guard was added to make it number[]

shamblonaut commented:
@pavish I implemented the changes to handle that scenario.

https://github.com/user-attachments/assets/511b5d97-a6c4-4f0e-a2b9-d1bebfdce5ed

Sorry for the late response.

pavish committed:
Merge branch 'develop' into feat-4806-data-explorer-duplicate

pavish reviewed (approved):
@shamblonaut Looks good to me. Thanks for the PR!
